### ISSUE TITLE
Toolshelf - Check if and how the other Ui code works in the toolshelf

### ISSUE THREAD

**Issue by ReinerBforartists** [Bforartists/Bforartists#1226]
We might need another buttonsize or a label or tabs or a panel or a menu. Let's check if there are ways to use them in the toolshelf.

**Comment by Draise14** [comment]
I have seen people add panels to the toolshelf with their addons. **Animation Nodes** is one example in the Node Editor.

**Comment by ReinerBforartists** [comment]
Thanks Draise. We will see :)

**Comment by ReinerBforartists** [comment]
Experiments time ...

**Comment by ReinerBforartists** [comment]
I can't even add a simple string between the buttons. And have no idea how we could add something like a tab functionality here -.-

**Comment by Draise14** [comment]
Yeah, the animation nodes only appends traditional panels, but doesn't place anything between buttons.

**Comment by ReinerBforartists** [comment]
Sad Reiner is sad ...

**Comment by ReinerBforartists** [comment]
Look what i just found ...

![hideheader](https://user-images.githubusercontent.com/13919702/74223972-219f4180-4cb8-11ea-9c7b-ed9f8c0d9e28.jpg)

**Comment by ReinerBforartists** [comment]
Would be nice to have the panel with a transparent background But the panel color is global. So one way would be to tweak the HIDE_HEADER option to show the header and just hide the background.

Keyword for further investigations: PNL_NO_HEADER

**Comment by Draise14** [comment]
This is wonderful news. Means.. there CAN be panels in the toolshelf.

**Comment by Draise14** [comment]
![bforartists_2020-02-11_13-08-28](https://user-images.githubusercontent.com/35357409/74264983-ae092e80-4ccf-11ea-994f-478e362d6303.png)
lol, this opens up so many doors. Panels! You can reorder and make and collapse groups of tools. I have seen this requested in other communities. The grey background is ok if it works like this.

**Comment by ReinerBforartists** [comment]
Yup. Now we just need the tabs back =3

**Comment by Draise14** [comment]
The tabs are implemented by C right? Sounds a bit trickier... But the nested buttons, panels and double row, I think solves, half the issues or keeping it all in a toolshelf.

**Comment by ReinerBforartists** [comment]
Yes, that's C code. Not sure if i am able to ressurect this one. From what i've read there is some incompatibility here. We will see :)

**Comment by ReinerBforartists** [comment]
The panel stuff makes me grey hairs. The spaghetti code goes everywhere. And dare you you miss one little bit. Can't get it to work so far. Not even simply duplicating an existing panel helps. -.-

**Comment by ReinerBforartists** [comment]
No way. With two panels i end in an internal error and an Internal Error crash. Seems that the system just allows one active panel in the tool shelf.

![interrnalerror](https://user-images.githubusercontent.com/13919702/75535802-9347f100-5a15-11ea-89e7-19e2acdab8d3.jpg)

**Comment by ReinerBforartists** [comment]
Okay, this makes even sense. The active tool can just be enumerated in one active panel. With two panels the whole system goes wonky therefore. When we want subpanels, then we would need to do that in the existing tool shelf panel. Not funny. And we need some kind of tabs for that, to turn things off and on.

What we can already do is to create a traditional panel in the tool shelf. This still works as usual. But that's outside of the tool shelf active tool system then. And we would need to develop a system to display text and icon buttons like in the rest of the tool shelf.

But you don't need a snap tool to stay active. It's a one trick tool, and performs its action with one click.

![snappanel](https://user-images.githubusercontent.com/13919702/75538939-c1c7cb00-5a19-11ea-9d04-30476191c204.jpg)

**Comment by ReinerBforartists** [comment]
Many thanks goes to @naetherm with his help to get the width functionality working!
![tswidth](https://user-images.githubusercontent.com/13919702/75603190-81755500-5acc-11ea-8d65-876b7c6d7ac8.jpg)

**Comment by ReinerBforartists** [comment]
Tabs are definitely removed. Let's dive into the log history ...

![notabs](https://user-images.githubusercontent.com/13919702/75603588-247b9e00-5ad0-11ea-97ac-4a1ef0239498.jpg)

**Comment by ReinerBforartists** [comment]
The commit that removed the tabs was:

Revision: c10f1379cdf28a66324a6f4db51b202014cf0015
Author: Campbell Barton <ideasman42@gmail.com>
Date: 10.06.2018 15:40:02
Message:
UI: remove tab use in the tool-settings

----
Modified: release/scripts/startup/bl_ui/space_view3d_toolbar.py
Modified: source/blender/editors/screen/area.c

Unfortunately, reverting the change in area.c does not lead to a solution. The error is still there. Investigating ...

**Comment by Draise14** [comment]
> 
> 
> No way. With two panels i end in an internal error and an Internal Error crash. Seems that the system just allows one active panel in the tool shelf.
> 
> ![interrnalerror](https://user-images.githubusercontent.com/13919702/75535802-9347f100-5a15-11ea-89e7-19e2acdab8d3.jpg)

You have to declare both panels at the footer with a unique name in the **space_toolsystem_toolbar.py**, just duplicating and listing the tools is not enough.. You also have to make sure the tools listed are... well set right? I get no errors when I have two or more panels with active tools. 

In the example below is a duplicate panel with a new name (TOOLS type of panel) and a hidden tool set as an example. 

![bforartists_2020-02-29_07-53-10](https://user-images.githubusercontent.com/35357409/75607802-ac5daa00-5ac8-11ea-8f02-b70d13c82c4d.png)
![bforartists_2020-02-29_07-52-54](https://user-images.githubusercontent.com/35357409/75607803-aec00400-5ac8-11ea-8256-0746d29f317f.png)

**Comment by ReinerBforartists** [comment]
Thanks Draise. I did. It was two different independant panels. But got the errors.

**Comment by Draise14** [comment]
![bforartists_2020-02-29_08-05-41](https://user-images.githubusercontent.com/35357409/75608078-6bff2b80-5aca-11ea-80da-27e6e3e9d39c.png)
Ok.. figured out how to make errors, but also how to not have errors. The little asterix * symbol pulls errors when placed on the CUT. But without the cut, it works fine with no errors. This might have a key. 

![bforartists_2020-02-29_08-06-27](https://user-images.githubusercontent.com/35357409/75608089-80432880-5aca-11ea-910b-f7b99d3b49ae.png)

**Comment by ReinerBforartists** [comment]
Cool. Thanks Draise. Then we can even split the tool shelf into smaller parts \o/

**Comment by Draise14** [comment]
I think the asterix is what defines the default active tool state, and if you have two different defaults in the same mode, it will throw an error. So the default should only be on one panel per mode, or be the same defaults as the original default if it's a duplicate. Did it work for you?

**Comment by ReinerBforartists** [comment]
Hm, thg

**Comment by Draise14** [comment]
Oops, closed by accident

**Comment by ReinerBforartists** [comment]
I haven't tested it yet. Will most probably touch it monday again. Sunday is merge ...

**Comment by ReinerBforartists** [comment]
Still no way. I run sooner or later into errors in the moment where i really split the tools apart.
Removing the asterisk will just make the sub menus inaccessible. It's needed for the tuple. So you did not fix it, you finally broke it ^^

Giving up. This is not really needed in case we can resurrect the tabs and put our other tools into those tabs with panels then. The ones in the text menu are all mostly one trick tools, and performs with one click. We don't need them in the tool shelf active tool chain. But you can do some further tests if you want.

Okay, on with the tabs. There must be a solution.

![noway](https://user-images.githubusercontent.com/13919702/75668356-81ff1e80-5c79-11ea-9203-56b6b5a332d3.jpg)

**Comment by Draise14** [comment]
You are not supposed to remove the asterixes. Do you want to see my sample file?



I swear there are no errors and it works fine.

**Comment by ReinerBforartists** [comment]
Yeah, that's the way it is now. 

As told, you can try if you can find a solution to that. I will go on with something that i am hopefully able to solve ^^

**Comment by ReinerBforartists** [comment]
Woah, searching for the missing bit ...

Next stop in the chain of finding out how to resurrect the tabs in the tool shelf is commit :

Revision: 6be7a98906b769c1496127552b524124a0e70c5a
Author: Campbell Barton <ideasman42@gmail.com>
Date: 26.10.2018 06:05:07
Message:
PyAPI: raise error when toolbar panels use tabs

Add-ons that register panels in the toolbar can
no longer use 'bl_categories' (tabs).
----
Modified: release/scripts/startup/bl_ui/space_toolsystem_toolbar.py
Modified: source/blender/editors/screen/area.c
Modified: source/blender/makesdna/DNA_screen_types.h
Modified: source/blender/makesrna/intern/rna_ui.c
Modified: source/blender/windowmanager/WM_toolsystem.h

**Comment by ReinerBforartists** [comment]
Bad news. Simply reverting the commit leads to a bunch of errors. Too much has changed since then. And there are too many loose ends. I am honestly lost with how this whole system works. It goes across dozens of files. And i haven't found the bit that resurrects the tabs yet. Waiting for help here.

**Comment by Draise14** [comment]
![bforartists_2020-03-05_09-50-15](https://user-images.githubusercontent.com/35357409/75993548-a0f2f000-5ec7-11ea-9343-01a5fb19f12b.png)
I've been thinking about the BFA1 buttons in the new toolshelf. 
- They would need to have align on True, as then the design would feel more consistent with the toolbar AND the new aligned buttons in the toolshelf
- I like the panels system
- I wonder if the new buttons could be the same size as the single use buttons in the other panels, and the padding aligned, then it COULD be a consistent. Icons could then be updated. 
- Tabs missing still......................  

![bforartists_2020-03-05_10-04-03](https://user-images.githubusercontent.com/35357409/75994707-5e321780-5ec9-11ea-8335-845537f6d914.png)
I then thought up that.. the top toolbar is good. But it could be better, maybe even it's own header WITHIN the 3D viewport (like a duplicate header, but only for buttons or toolbar code). Still no tabs though.... 

Then with this video, I saw someone make a workign prototype of an editor switch from a toolbar [here](https://www.youtube.com/watch?v=ZphqykbmHeQ&feature=youtu.be):
![firefox_2020-03-05_10-15-36](https://user-images.githubusercontent.com/35357409/75995394-5a52c500-5eca-11ea-87ec-16999eb79d6c.png)

What if that same system could be implemented in the 3D viewport to toggle header one click button ribbons like our toolbar - or appended/remove panels (or collapse panels) in the toolbar with BFA1 style buttons ribbons?

**Comment by ReinerBforartists** [comment]
That's far beyond any scope of what i am able and willing to do , sorry. But the asset manager will come with Blender 2.83 anyways, in the one or another form.

The big icons is part of the new tool system. You cannot use it at the old panels this easily. And i find the idea to have the smaller buttons to indicate that this is the old tool system much more appealing. But for that we need tabs ...

**Comment by ReinerBforartists** [comment]
Decision time. Readding tabs turned out to be unmanageable for us. They have changed the method. I have still no idea how this all is meant to work. And reverting the patches did not help. We might reinvestigate into this at one point. I will make a task for it now. But the gold version of Bforartists 2 will come without them for sure.

We have now explored what is possible and what not. The rest can go into single issues. And this means this issue has fulfilled its purpose.

Closing.
